packages/sources/coinbase-prime/src/transport/balance.ts

```diff
@@ -0,0 +1,101 @@
+import { HttpTransport } from '@chainlink/external-adapter-framework/transports'
+import { BaseEndpointTypes } from '../endpoint/balance'
+import { sign } from './utils'
+
+export interface ResponseSchema {
+  balances: {
+    symbol: string
+    amount: string
+    holds: string
+    bonded_amount: string
+    reserved_amount: string
+    unbonding_amount: string
+    unvested_amount: string
+    pending_rewards_amount: string
+    past_rewards_amount: string
+    bondable_amount: string
+    withdrawable_amount: string
+    fiat_amount: string
+  }[]
+  type: string
+  trading_balances: {
+    total: string // Returns total in fiat amount
+    holds: string
+  }
+  vault_balances: {
+    total: string // Returns total in fiat amount
+    holds: string
+  }
+}
+
+export type HttpTransportTypes = BaseEndpointTypes & {
+  Provider: {
+    RequestBody: never
+    ResponseBody: ResponseSchema
+  }
+}
+export const httpTransport = new HttpTransport<HttpTransportTypes>({
+  prepareRequests: (params, config) => {
+    return params.map((param) => {
+      const timestamp = Math.floor(Date.now() / 1000)
+      const method = 'GET'
+      const path = `/v1/portfolios/${param.portfolio}/balances`
+      const message = `${timestamp}${method}${path}`
+      const signature = sign(message, config.SIGNING_KEY)
+      return {
+        params: [param],
+        request: {
+          baseURL: config.API_ENDPOINT,
+          url: path,
+          headers: {
+            'X-CB-ACCESS-KEY': config.ACCESS_KEY,
+            'X-CB-ACCESS-PASSPHRASE': config.PASSPHRASE,
+            'X-CB-ACCESS-SIGNATURE': signature,
+            'X-CB-ACCESS-TIMESTAMP': timestamp,
+            'Content-Type': 'application/json',
+          },
+          params: {
+            symbols: param.symbol.toUpperCase(),
+            balance_type: `${param.type.toUpperCase()}_BALANCES`,
+          },
+        },
+      }
+    })
+  },
+  parseResponse: (params, response) => {
+    // The adapter only supports querying one asset at a time so the balances list should only contain 1 element
+    if (!response.data || !response.data.balances || response.data.balances.length !== 1) {
+      return params.map((param) => {
+        return {
+          params: param,
+          response: {
+            errorMessage: `The data provider did not return a proper response for Portfolio: ${param.portfolio}, Balance Type: ${param.type}, Symbol: ${param.symbol}`,
+            statusCode: 502,
+          },
+        }
+      })
+    }
```

Comment:
Would it be too much to split out these error cases? It may be useful if the errorMessage specifies `data.balances not present` or `data.balances.length is not 1`

Reply:
Good call! I've updated the code in the latest commit. I also cleaned up all of the `params.map` loops and moved all of the logic into one. The way the code was set up previously doesn't really optimize anything since we're not batching.

Reply:
Cool, I'll approve then